### PR TITLE
Add support for libssh connection type

### PR DESCRIPTION
**Pull request by ganeshrn**
##### SUMMARY

*  Add libssh connection type that uses
   pylibssh library (https://github.com/ansible/pylibssh)
   which is Python extension for libssh (https://www.libssh.org/)
*  Refactor network_cli connection plugin to work with both
   libssh and paramiko_ssh connection type under the hood.
   The default connection type is set to libssh and can be
   configured using `ssh_transport_type` network_cli
   plugin configuration.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

plugins/connection/libssh.py
plugins/connection/network_cli.py

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,3 +2,4 @@ ansible
 jxmlease
 ncclient
 paramiko
+ansible-pylibssh==0.0.1a2.dev0
```
Comment (webknjaz) [inline requirements.txt:5]: This may require `--pre`

**Comment by pabelanger** [comment]
recheck

**Comment by opoplawski** [comment]
Very interested in this, thanks.  Trying to connect to an ios device fails with become: enable
```
2020-07-01 16:17:12,995 p=21896 u=myuser n=ansible |  21896 1593641832.99550: Loading ShellModule 'sh' from /export/home/myuser/ansible/lib/ansible/plugins/shell/sh.py (found_in_cache=True, class_only=False)
2020-07-01 16:17:12,995 p=21896 u=myuser n=ansible | <sw-cisco-2960s-2.> ESTABLISH LIBSSH CONNECTION FOR USER: None on PORT 22 TO sw-cisco-2960s-2.
2020-07-01 16:17:14,648 p=21896 u=myuser n=ansible | ssh connection is OK: <pylibsshext.session.Session object at 0x7f6f1a61f290>
2020-07-01 16:17:15,423 p=21896 u=myuser n=ansible | Traceback (most recent call last):
  File "/export/home/myuser/ansible/lib/ansible/utils/jsonrpc.py", line 46, in handle_request
    result = rpc_method(*args, **kwargs)
  File "/export/home/myuser/.ansible/collections/ansible_collections/ansible/netcommon/plugins/connection/network_cli.py", line 480, in update_play_context
    self._terminal.on_become(passwd=auth_pass)
  File "/export/home/myuser/.ansible/collections/ansible_collections/cisco/ios/plugins/terminal/ios.py", line 101, in on_become
    % (prompt, e.message)
AnsibleConnectionFailure: unable to elevate privilege to enable mode, at prompt [None] with error: For matched prompt '
Password: ', answer is not valid
```
this is with latest ansible git on EL 7.8 w/ python 2.7

**Comment by ganeshrn** [comment]
@opoplawski Will need more information on your environment.
What is the value of ansible_connection set.
To work with this PR you need to install ansible-pylibssh library and
set variables:

ansible_connection: network_cli
ansible_network_cli_ssh_type: libssh

After running playbook you should see the log
ssh_type set to libssh

**Comment by opoplawski** [comment]
Yes, ansible-pylibssh is installed and I'm using network_cli and libssh.  It works if I set ansible_network_cli_ssh_type to paramiko:
```
2020-07-02 10:49:28,050 p=7160 u=myuser n=ansible | response-2: 
sw-cisco-2960s-2#
2020-07-02 10:49:28,050 p=7160 u=myuser n=ansible | matched cli prompt '
sw-cisco-2960s-2#' with regex '[\r\n]?[\w\+\-\.:\/\[\]]+(?:\([^\)]+\)){0,3}(?:[>#]) ?$' from response 'enable
Password: 
sw-cisco-2960s-2#'
```

**Comment by ganeshrn** [comment]
@opoplawski  I have pushed a commit to the PR to fix the enable prompt issue. Can you please fetch the latest changes from the PR and test it.

**Comment by opoplawski** [comment]
@ganeshrn thanks, that looks better - enable appears to work now.

Next issue I'm seeing appears when using --diff, though I think it would also show as incorrect changed status as well.  There appears to be some kind of buffering issue:
```
--- before
+++ after
@@ -1,4 +1,4 @@
-sshow runrunning-ng-confifig
+shshow runningng-cononfonfig
```
similar can also be seen in the responses logged:
```
2020-07-06 10:18:37,546 p=30131 u=myuser n=ansible | send command: terminal length 0
2020-07-06 10:18:37,546 p=30131 u=myuser n=ansible | command: terminal length 0
2020-07-06 10:18:37,547 p=30131 u=myuser n=ansible | response-1: tertermi
2020-07-06 10:18:37,547 p=30131 u=myuser n=ansible | response-2: nal lnal leng
2020-07-06 10:18:37,547 p=30131 u=myuser n=ansible | response-3: th
2020-07-06 10:18:37,547 p=30131 u=myuser n=ansible | response-4: 0
0
sw-cisco-2960s-2>
```
compared to when running with paramiko:
```
2020-07-06 10:28:46,286 p=32152 u=myuser n=ansible | send command: terminal length 0
2020-07-06 10:28:46,286 p=32152 u=myuser n=ansible | command: terminal length 0
2020-07-06 10:28:46,286 p=32152 u=myuser n=ansible | response-1: terminal length
2020-07-06 10:28:46,286 p=32152 u=myuser n=ansible | response-2: 0

2020-07-06 10:28:46,286 p=32152 u=myuser n=ansible | response-3: sw-cisco-2960s-2>
```

**Comment by ganeshrn** [comment]
@opoplawski https://github.com/ansible/pylibssh/pull/89 should hopefully fix the buffer issue.
You can apply the patch locally or install it using pip from nightly build here https://ansible.github.io/pylibssh/simple/ansible-pylibssh/ after it is merged

**Comment by opoplawski** [comment]
Looks like it is working now.  Thanks!

**Comment by pabelanger** [comment]
I believe this is going to break jobs, see: https://e124d40ad7e5a4cfce03-64465fc08042d647cf2bd0dae766d04f.ssl.cf1.rackcdn.com/72/99eec5c843f3081495a6a3063fdaa935bbf817bf/gate/ansible-test-network-integration-eos-httpapi-python27-stable29/eb21369/job-output.html#l1585

I'm going to dequeue from gate, for the moment until we can review.

**Comment by ganeshrn** [comment]
@pabelanger Test are passing now https://github.com/ansible-collections/arista.eos/pull/87

**Comment by jean-christophe-manciot** [comment]
@ganeshrn 
Do you plan to update [network_cli documentation](https://docs.ansible.com/ansible/latest/plugins/connection/network_cli.html) to incorporate ```ssh_transport_type``` parameter?
Also, it would be very handy to define a behavioral inventory parameter, such as ```ansible_ssh_transport_type```.

**Pull request merged**